### DISCUSSION TITLE
Custom Views / Component needs to be the next big focus of Payload.

### DISCUSSION THREAD

**Discussion by tyteen4a03** [payloadcms/payload#3960] [General]
Payload is great as a headless CMS - no questions about that. But most people need more than what the stock Admin UI provides, and right now, putting in custom views is an uphill battle.

I've been trying to put together a simple custom view and I ran into many problems:
* No documentation on _authoring_ custom views - sure, the documentation tells you how to put one in, but I have to either look at source code or watch somebody else's videos to figure out where the building blocks are, nor does it have good tutorials on how to put custom views together beyond "here's a page". (The thing is once you find those components, like `<RenderFields>`, you realize you might not need to reinvent as many wheels)
* Types don't match between custom components and stock components, which makes source code reading a mental gymnastics exercise as you try to map out how to make the stock code work for your custom components
* No searchable API documentation
* No design system documentation / storybook (although I know the team is working on this)
* Design system is too monochrome - colors are hard to come by which makes distinguishing elements hard for users
* Custom views cannot modify `<head>` (this might be fixed when we move to Next?)
* No testing guides on custom components / views
* Community help Discord channel largely ignored - I know Payload team is small and I know the team commits every single day but it really does seem like most of the community help posts get no attention.

I think it's fair to say most people need to customize the Admin UI beyond what's provided in stock every single day, and right now the development experience is far from great. I really hope the team makes this their next big focus because this is the last 10% that would make Payload a truly killer CMS.

**Comment by paulpopus** [comment]
I read everything you wrote and I agree, having spoken with members of the team in various occasions I can also tell you they're very much aware of these pain points.

> No searchable API documentation

Can you expand on this point for me please? What exactly are you missing or looking for when you mean searchable API documentation

**Reply by tyteen4a03** [reply]
Everything that is exposed by Payload (whether local API or components) should be documented online with functionality description, parameter descriptions, and usage examples. Right now Payload doesn't expose this information on the web and navigating TypeScript definitions isn't fun.

For example: it took me a lot of source code reading (where I had to clone the payload source repo) to find out what `fieldTypes` of `<RenderFields>` is supposed to be.

**Comment by yonnic** [comment]
exactly one year later and all the mentioned pain points remain in the new 3.0 beta
everything related to the UI is just a black box and ultra difficult to figure out. Looks like its faster to come up with a fully custom designed and coded custom view replacing the whole UI page and not even try to reuse the native UI elements and layout at all

**Reply by revnelson** [reply]
There are definitely improvements with the new 3.0 beta, but I agree that there is a lot of room for improvement that should be considered high priority. I created a [discussion](https://github.com/payloadcms/payload/discussions/8928) recently about the RelationshipSelect field not being usably exposed, and there are others too.

Each field type should have an exposed `Input` component that can be used in any custom views. The currently exposed inputs are `TextInput` `TextAreaInput` `CheckboxInput` `SelectInput` and `UploadInput`.

In the event that we want to completely roll our own input, helper functions used internally should also be exposed. In the case of building out a RelationshipInput, for example, utils such as `createRelationMap` `findOptionsByValue` and `optionsReducer` used by the `RelationshipField` component would need to be exposed.

Lastly, passing in types to some of the core functions should be expanded. One example is the `admin.hidden` function which receives a `user` prop defined as the internal Payload `User` type. However, our generated `User` type is likely different with additional fields that may be used in such a function, such as `roles`. This leads to clumsy typing such as:
```ts
user as unknown as User
```

**Reply by nachomglz** [reply]
This is a fact, I also find it a big pain every time I want to create a custom field, you usually end up going for a fully custom solution as there's no documentation whatsoever on how things work, and it takes less time to code something  from scratch than trying to understand how the components work going through TS definitions or cloning the source code.

I think there's huge room for improvement in this aspect.
